### ISSUE TITLE
localStorage stub doesn't work on 1.17.3

### ISSUE THREAD

**Issue by ilgooz** [sinonjs/sinon#1046]
in version 1.12.2 everything is good on Chrome http://jsbin.com/noqigilome/edit?html,js,console,output
but on 1.17.3 it doesn't seem to work http://jsbin.com/fufoyomuja/edit?html,js,output

ubuntu 14.10 - Chrome 48.0.2564.109 (64-bit)

thank you.

```
"TypeError: Attempted to wrap string property getItem as function
    at checkWrappedMethod (http://sinonjs.org/releases/sinon-1.17.3.js:1355:29)
    at Object.wrapMethod (http://sinonjs.org/releases/sinon-1.17.3.js:1398:21)
    at Object.stub (http://sinonjs.org/releases/sinon-1.17.3.js:3465:26)
    at qulovapigu.js:1:12
    at https://static.jsbin.com/js/prod/runner-3.35.12.min.js:1:13891
    at https://static.jsbin.com/js/prod/runner-3.35.12.min.js:1:10820"
```

**Comment by fatso83** [comment]
Thanks for reporting this. Seems to be a bug, but not sure what causes it after inspecting the source.

Adding output to your report.

**Comment by fatso83** [comment]
@ilgooz Would you mind trying [my approach](https://github.com/sinonjs/sinon/issues/821#issuecomment-139280107) of finding where this regression was introduced? You just need to modify the [test script](https://raw.githubusercontent.com/fatso83/Sinon.JS/setImmediate-bug/test-script.sh) a little bit to make git do all the work for you in finding where this bug was introduced.

It is a quite fun exercise and you could reuse it in other projects :)

**Comment by jonnyreeves** [comment]
@ilgooz please could you provide an example which reproduces this issue?  Thanks.

**Comment by fatso83** [comment]
@jonnyreeves he has :-) He just used JSBin to make it runnable in the browser, as a simple Node test would not show it. I verified it. 

I can perhaps try to follow my own instructions above to find where it was introduced.

**Comment by jonnyreeves** [comment]
@fatso83 @ilgooz yep sorry, saw the jsbin example just after commenting!  I've since deleted my comment, I'll delete your and this one as well to keep the log for this issue on topic - sorry for the spam.

**Comment by fatso83** [comment]
I modified the mentioned test script to use a test case that shows this issue ([view it here](https://gist.github.com/fatso83/7f135d2de168a07582b075b0fa46e9a9)). `git bisect` found the regression to have been introduced in bb467e6 (PR #692 Feature: getters and setters).

Maybe @simonzack can shed some light on this? 
<details>
<summary>Test case from the test script</summary>
describe('#1046', function() {

    afterEach(function(){
        try{
        localStorage.getItem.restore();
        }catch(e){}
    });

    function aTest() {
        sinon.stub(localStorage, 'getItem', function() {
              return "you see me?";
        });

        localStorage.getItem('foo');
    }

    it('should not throw (#1)', aTest);
    it('should not throw (#2)', aTest);
});
</details>

Basically, `stub.restore()` is not working as it should, as the stubbing function thinks `localStorage.getItem` has become a string property the second time around.

**Comment by fatso83** [comment]
OK, the problem seems to be that when restoring localStorage a property (or a value in the store - not sure which) called `getItem` with the value `spy` is being added to localStorage. You then end up with both a function called `getItem` and a property called `getItem` - leading to the bug. I am trying to debug why.

**Comment by fatso83** [comment]
Found the issue and a fix is supplied as #1098.

**Comment by fatso83** [comment]
Fix or no fix, trying to stub native objects is fragile and not something that usually works across browsers. Even with the fix, or in v1.12, this does not work in Firefox (and never worked).

You are much better off with some kind of dependency injection (see this [SO answer](http://stackoverflow.com/a/20153543/200987) for a good example).

**Comment by fatso83** [comment]
This has been merged into the sinon#v1.17. A fix should also be supplied to the `master` branch.

**Comment by fatso83** [comment]
Fixed in `master` as #1111
